operator/builtin/input/file/design.md

```diff
@@ -0,0 +1,146 @@
+        - In the vast majority of cases, this occurs during file rotation that uses the copy/truncate method. (See fingerprinting section above.)
+6. Reader Creation
+    1. Each file handle is wrapped into a `Reader` along with some metadata. (See Reader section above)
+        - During the creation of a `Reader`, the file's fingerprint is crossreferenced with previously known fingerprints.
+        - If a file's fingerprint matches one that has recently been seen, then metadata is copied over from the previous iteration of the Reader. Most importantly, the offset is accurately maintained in this way.
+        - If a file's fingerprint does not match any recently seen files, then its offset is initialized according to the `start_at` setting.
+7. Detection of "Lost" Files
+8. Consumption
+    1. Lost files are consumed. Since these files are thought to have been rotated to a location where we will not match them again, we will never see them again. 
+        - The best we can do is finish consuming their current contents.
+        - We can reasonable expect in most cases that these files are no longer being written to.
+    2. Matched files (from this poll cycle) are consumed.
+        - These file handles will be left open until the end of the next poll cycle.
+        - Typically, we can expect to find most of these files again. However, these files are consumed greedily, in case we do not see them again.
+    3. All files in both sets are consumed concurrently.
+9. Closing
+    1. All files from the previous poll cycle are closed.
+10. Archiving
+    1. Readers created in the current poll cycle are added to the historical record.
+11. Pruning
+    1. The historical record is purged of Readers that have existsed for 3 generations.
+        - This number is somewhat arbitrary, and should probably be made configurable. However, its exact purpose is quite obscure.
+12. Persistence
+    1. The historical record of readers is synced to whatever persistence mechanism was provided to the operator.
+13. End Poll Cycle
+    1. At this point, the operator sits idle until the poll timer fires again.
+14. Start Poll Cycle
+15. Dequeueing
+    1. If any matches are queued from the previous cycle, an appropriate number are dequeued, and processed that same as would a newly matched set of files.
+16. Aging
+    1. If no queued files were left over from the previous cycle, then all previously matched files have been consumed, and we are ready to query the file system again. Prior to doing so, we will increment the "generation" of all historical Readers.
+
+
+# Additional Details
+
+### Startup Logic
+
+Whenever the operator starts, it:
+- Requests the historical record of Readers, as described in steps 10-12 of the poll cycle.
+- Starts the polling timer.
+
+### Shutdown Logic
+
+When the operator shuts down, it:
```

Comment:
Do we shutdown midcycle or always wait for the cycle to complete? Is there a chance to store the checkpoint before the data is actually consumed and shutdown happens?

Reply:
I'll update the doc, but to answer your question: 

If a poll cycle is _not_ currently underway, the operator simply closes any open files. 

Otherwise, the current poll cycle is signaled to stop immediately, which in turn signals all Readers to stop immediately.

If a Reader is idle or in between log entries, it will return immediately. Otherwise is will return after consuming one final log entry.

When all Readers return, the remainder of the poll cycle completes as usual, which includes the steps labeled Closing, Archiving, Pruning, and Persistence.

The net effect of all this is that all files are checkpointed in a normal manner (i.e. not in the middle of a log entry), and all checkpoints are persisted.